### PR TITLE
feat: set up for migration to effect sets

### PR DESCRIPTION
**Pull request by mlutze**
related to #3542. This only contains some unused Parser structures. I am separating this from the next step since there may be complications to handle in the Weeder in order to support the new grammar without breaking the old.

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 1]
```diff
@@ -1480,6 +1480,119 @@ object ParsedAst {
...
   }
 
+  /**
+    * Effect Set
+    */
+  sealed trait EffectSet
+
+  object EffectSet {
+    /**
+      * Singleton effect set.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1490]: Represents an effect set with a single effect.

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 2]
```diff
@@ -1480,6 +1480,119 @@ object ParsedAst {
...
+      *
+      * @param sp1 the position of the first character in the set.
+      * @param eff the effect.
+      * @param sp2 the position of the last character in the set.
+      */
+    case class Singleton(sp1: SourcePosition, eff: Effect, sp2: SourcePosition) extends EffectSet
+
+    /**
+      * Pure effect set.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1499]: Represents the empty effect set. Written as `{ }` or `{ Pure }` depending on the syntactic context.

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 3]
```diff
@@ -1480,6 +1480,119 @@ object ParsedAst {
...
+      *
+      * @param sp1 the position of the first character in the set.
+      * @param sp2 the position of the last character in the set.
+      */
+    case class Pure(sp1: SourcePosition, sp2: SourcePosition) extends EffectSet
+
+    /**
+      * A set of effects.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1507]: Represents a set of effects.

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 4]
```diff
@@ -1480,6 +1480,119 @@ object ParsedAst {
...
+      * @param sp2 the position of the last character in the effect.
+      */
+    case class Complement(sp1: SourcePosition, eff: ParsedAst.Effect, sp2: SourcePosition) extends ParsedAst.Effect
+
+    /**
+      * Represents a read of the region variables `regs`.
+      *
+      * @param regs the region variables that are read.
+      */
+    case class Read(sp1: SourcePosition, regs: Seq[Name.Ident], sp2: SourcePosition) extends ParsedAst.Effect
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1569]: These should go right after Var. Usually the order is:

- Var like things
- Constant like things
- Compound things.

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 5]
```diff
@@ -1480,6 +1480,119 @@ object ParsedAst {
...
+      * A reference to an declared effect.
+      *
+      * @param sp1  the position of the first character in the effect.
+      * @param name the declared effect.
+      * @param sp2  the position of the last character in the effect.
+      */
+    case class Eff(sp1: SourcePosition, name: Name.QName, sp2: SourcePosition) extends ParsedAst.Effect
+
+    /**
+      * The Impure effect.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1589]: Missing line break.

Maybe some documentation about how this is used?

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 6]
```diff
@@ -1480,6 +1480,119 @@ object ParsedAst {
...
+      * @param sp1  the position of the first character in the effect.
+      * @param sp2  the position of the last character in the effect.
+      */
+    case class Impure(sp1: SourcePosition, sp2: SourcePosition) extends ParsedAst.Effect
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1593]: This should go after Read, Write, Impure.

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 7]
```diff
@@ -1480,6 +1480,119 @@ object ParsedAst {
...
+      * @param regs the region variables that are written.
+      */
+
+    case class Write(sp1: SourcePosition, regs: Seq[Name.Ident], sp2: SourcePosition) extends ParsedAst.Effect
+
+    /**
+      * A reference to an declared effect.
+      *
+      * @param sp1  the position of the first character in the effect.
+      * @param name the declared effect.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1583]: the fully-qualified name of the effect?

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 8]
```diff
@@ -1480,6 +1480,119 @@ object ParsedAst {
...
+      * @param regs the region variables that are read.
+      */
+    case class Read(sp1: SourcePosition, regs: Seq[Name.Ident], sp2: SourcePosition) extends ParsedAst.Effect
+
+    /**
+      * Represents a write of the region variables `regs`.
+      *
+      * @param regs the region variables that are written.
+      */
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1576]: spurious line break

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 9]
```diff
@@ -1480,6 +1480,119 @@ object ParsedAst {
...
+    /**
+      * Effect variable.
+      *
+      * @param sp1   the position of the first character in the effect.
+      * @param ident the name of the variable.
+      * @param sp2   the position of the last character in the effect.
+      */
+    case class Var(sp1: SourcePosition, ident: Name.Ident, sp2: SourcePosition) extends ParsedAst.Effect
+
+    /**
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1531]: How about these names and in this order:

- Complement
- Union
- Intersection
- Difference

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala [hunk 10]
```diff
@@ -1393,6 +1399,86 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
...
+    }
+
+    def Complement: Rule1[ParsedAst.Effect] = rule {
+      SP ~ operatorX("~") ~ optWS ~ SimpleEffect ~ SP ~> ParsedAst.Effect.Complement
+    }
+
+    def Parens: Rule1[ParsedAst.Effect] = rule {
+      "(" ~ Effect ~ ")"
+    }
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Parser.scala:1452]: The above should be re-ordered to approximately match the declarations.

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala [hunk 11]
```diff
@@ -1393,6 +1399,86 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
...
+      operatorX("-") ~ oneOrMore(SimpleEffect).separatedBy(optWS ~ "-" ~ optWS) ~> ParsedAst.Effect.Minus
+    }
+
+    def IntersectTail = rule {
+        operatorX("&") ~ oneOrMore(SimpleEffect).separatedBy(optWS ~ "&" ~ optWS) ~> ParsedAst.Effect.Intersect
+    }
+
+    def BinaryTail = rule {
+      PlusTail | MinusTail | IntersectTail
+    }
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Parser.scala:1479]: We dont use this style in other parts of the parser. What explains the difference?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:1479]: Is it because you dont want to backtrack over the first simple effect?

**Comment by magnus-madsen** [comment]
Overall looks good. Many proposals for minor tweaks.

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala (hunk 11 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:1479]: Yes, that plus the decision to require parentheses if it's not a chain of the same operator.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:1479]: OK

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 12]
```diff
@@ -1480,6 +1480,120 @@ object ParsedAst {
...
   }
 
+  /**
+    * Effect Set
+    */
+  sealed trait EffectSet
+
+  object EffectSet {
+    /**
+      * Represents and effect set with a single effect.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1490]: an

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1490]: Ill merge it as is 👍

**Pull request merged**